File: src/main/java/station/SecretStation.java
package station;

import secret.Secret;
import secret.SecretPart;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

import static secret.Secret.generateSecret;
import static util.Util.matrixMultiplication;

public class SecretStation extends Station {
    private Secret secret;

    public void shareSecret(List<DataStation> parties) {
        this.secret = generateSecret(parties);
        for (DataStation station : parties) {
            for (SecretPart part : secret.getParts()) {
                if (part.getId().equals(station.getId())) {
                    station.setLocalSecret(part);
                }
            }
        }
    }

    public DataStation generateDataStation(List<String> subset) {
        List<BigInteger[][]> list = new ArrayList<>();
        for (String id : subset) {
            for (SecretPart part : secret.getParts()) {
                if (part.getId().equals(id)) {
                    list.add(part.getMatrix());
                }
            }
        }
        String id = subset.stream().map(e -> e.toString()).reduce("", String::concat);
        return new DataStation(id, matrixMultiplication(list));

    }


}

File: python/station/SecretStation.py
from secret import Secret
from secret import generateSecret
from station import DataStation
from util.Util import matrixMultiplication


class SecretStation:
    def __init__(self):
        self.__secret = None

    def shareSecret(self, parties:[DataStation]):
        self.__secret = generateSecret(parties)
        for party in parties:
            for secretPart in self.__secret.getParts():
                if secretPart.getId() == party.getId():
                    party.setLocalSecret(secretPart)

    def generateDataStation(self, subset: [str]):
        list = []
        for id  in subset:
            for part in self.__secret.getParts():
                if part.getId() == id:
                    list.append(part.getMatrix())

        return DataStation.DataStation(''.join(subset), matrixMultiplication(list))